src/ar_http_iface_h.erl

```diff
@@ -8,22 +8,43 @@
 %%%===================================================================
 
 init(InitialReq, State) ->
-	{Status, Headers, Body, HandledReq} = do_handle(InitialReq),
+	TimeoutPid = handler_timeout_after(InitialReq, 30000),
+	{Status, Headers, Body, HandledReq} = handle(InitialReq),
 	CowboyStatus = handle208(Status),
 	RepliedReq = cowboy_req:reply(CowboyStatus, Headers, Body, HandledReq),
+	TimeoutPid ! cancel,
 	{ok, RepliedReq, State}.
 
 %%%===================================================================
 %%% Private functions.
 %%%===================================================================
 
-do_handle(Req) ->
+%% @doc Because cowboy 2.0 seemingly doesn't support accessing the request from
+%% multiple processes (see https://github.com/ninenines/cowboy/issues/1374) the
+%% following is the only way to introduce a asynchronous handler timeout.
+handler_timeout_after(Req, T) ->
+	HandlerPid = self(),
+	spawn_link(fun() ->
+		receive
+			cancel -> ok
+		after
+			T ->
+				ar:warn([
+					handler_timeout,
+					{method, cowboy_req:method(Req)},
+					{path, cowboy_req:path(Req)}
+				]),
+				exit(HandlerPid, kill)
```

Comment:
It seems like the HTTP request hangs when this happens. I would expect Cowboy to either return a 500 or at least cut the connection. I don't think this is ok.